docs/spec/governance/software_upgrade.md

```diff
@@ -0,0 +1,133 @@
+* Current version commit hash: [XXX]
+* New version Github URL: https://github.com/...
+* New version: [XXX]
+* New version commit hash: [XXX]
+* Upgrade prepared by:
+ - <name/handle> on <date>; commit signature: https://github.com/...
+* Audited by:
+ - <name/handle> on <date>; report: https://github.com/...
+ - <name/handle> on <date>; report: https://github.com/...
+
+## Description:
+
+Describe the purpose of the upgrade and other details.
+
+- [ ] Link to any previous discussions
+- [ ] Link to any alternative proposals
+
+## Requirements:
+
+Describe any requirements
+
+- [ ] Any new hardware requirements?
+- [ ] Any new bandwidth requirements?
+- [ ] Any dependencies on other proposals?
+
+## Procedure:
+
+* Deadline: By 2 weeks from after the proposal has been accepted.
+
+Once this proposal has been accepted, validators should upgrade to the new
+software as described above by the deadline.  Once the new software is
+deployed, the new software will automaticaly submit a message to the
+SoftwareUpgradeKeeper.  The SoftwareUpgradeKeeper ensures that a sufficient
+quorum of validators (SOFTWARE_UPGRADE_QUORUM parameter) is running a specific
+version of the blockchain software.
+
+Once a sufficient quorum has been reached, the "SOFTWARE_UPGRADE_[XXX]_ENABLED"
+parameter will get set to true, thus switching on the new logic.
+
+## Conflict Resolution
+
+In case there are conflicting proposals, the most recent proposal to get
+accepted should be followed.  In other words, if another proposal gets accepted
+after this proposal does which conflicts with this proposal, the new proposal
+should be heeded instead of the process declared here.
+```
+
+It is recommended that all links point to Github for now, for accountability.
+
+This proposal template assumes that much discussion about the upgrade has
+already taken place elsewhere.  To gauge interest in a proposed upgrade,
+users should instead submit a survey proposal with the following template:
+
+Notethat the `Procedure` section has already been filled out.  Do not change
+this section unless there is a good reason to do so.
+
+```
+# SOFTWARE UPGRADE SURVEY
+
+THIS IS A NONBINDING SURVEY.  VOTES HERE ARE ONLY INTENDED TO GUAGE INTEREST.
+IF THIS PROPOSAL MANDATES ANY ACTION, IT IS THE VOTER'S DUTY TO VOTE "NO".
+
+<fill out as much as possible from the SOFTWARE UPGRADE PROPOSAL template>
+```
+
+### Step 2: Validators upgrade their software
+
+Once the proposal from step 1 has been accepted, validators should upgrade
+their software.  For the standard upgrade procedure, it is OK for non-validator
+full nodes to upgrade as well.
+
+### Step 3: Update the SoftwareUpgradeKeeper
+
+The new version of the software should automatically include a (single) signed
+message to the SoftwareUpgradeKeeper w/ the new version of software that the
+validator is running.  Even if a validator were to change their software
+version multiple times, the SoftwareUpgradeKeeper ensures that nothing happens
```

Comment:
I guess the keeper also ensures that validator can only signal a single version to a keeper? (1 to 1 map)